x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java

```diff
@@ -382,6 +383,22 @@ protected LogicalPlan rule(Limit limit) {
                 if (unary instanceof Eval || unary instanceof Project || unary instanceof RegexExtract || unary instanceof Enrich) {
                     return unary.replaceChild(limit.replaceChild(unary.child()));
                 }
+                // limit after mv_expand or limit and sort after mv_expand (| mv_expand x | sort y | limit 5)
+                else if (unary instanceof MvExpand || (unary instanceof OrderBy orderBy && orderBy.child() instanceof MvExpand)) {
+                    MvExpand mvExpand = unary instanceof MvExpand mve ? mve : (MvExpand) (unary.child());
+                    Limit limitBeforeMvExpand = limitBeforeMvExpand(mvExpand);
+                    // if there is no "appropriate" limit before mv_expand, then push down a copy of the one after it so that:
+                    // - a possible TopN is properly built as low as possible in the tree (closed to Lucene)
+                    // - the input of mv_expand is as small as possible before it is expanded (less rows to inflate and occupy memory)
+                    if (limitBeforeMvExpand == null) {
+                        var duplicateLimit = new Limit(limit.source(), limit.limit(), mvExpand.child());
+                        if (unary instanceof OrderBy orderBy) {
+                            return limit.replaceChild(orderBy.replaceChild(mvExpand.replaceChild(duplicateLimit)));
+                        } else {
+                            return limit.replaceChild(mvExpand.replaceChild(duplicateLimit));
+                        }
+                    }
+                }
```

Comment:
This block needs to be simplified as the main condition is checked 2-3 times.
do a simple boolean isMvExpand = unary instanceof MvExpand, boolean orderByChild = unary instanceof OrderBy && ...
then use those later in the code.